### PR TITLE
Add a convenience function for calling the builtin `hash`

### PR DESCRIPTION
**Pull request by tjstum**
pybind11 allows for calling `pybind11::hash(foo)` in C++ as an equivalent of `hash(foo)` in Python. This is handy when binding a type whose Python hash should match that of some other Python object (like a string, perhaps). Expose something similar for nanobind

Test included

### CODE REVIEW

File: include/nanobind/nb_lib.h [hunk 1]
```diff
@@ -154,6 +154,10 @@ NB_CORE size_t obj_len_hint(PyObject *o) noexcept;
 /// Obtain a string representation of a Python object
 NB_CORE PyObject* obj_repr(PyObject *o);
 
+/// Obtain a hash value of a Python object
+NB_CORE Py_hash_t obj_hash(PyObject *o);
+
```
Comment (wjakob) [inline include/nanobind/nb_lib.h:159]: remove newline

File: tests/test_classes.cpp [hunk 2]
```diff
@@ -435,4 +435,13 @@ NB_MODULE(test_classes_ext, m) {
 
     using NonCopyableVec = std::vector<NonCopyable>;
     nb::class_<NonCopyableVec>(m, "NonCopyableVec");
+
+    // test32_deference
+    struct Deferential {nb::object target; };
```
Comment (wjakob) [inline tests/test_classes.cpp:440]: space after `{`.

Can you rename the class from `Dereferential` (which I think is more common to express "honor" to another person) to something like `Indirect` or `HashWrapper`? Same for the test case name.

File: tests/test_classes.cpp [hunk 3]
```diff
@@ -435,4 +435,13 @@ NB_MODULE(test_classes_ext, m) {
...
+    nb::class_<Deferential>(m, "Deferential")
+        .def(nb::init<nb::object>())
+        .def("__eq__", [](const Deferential &self, nb::object other){
```
Comment (wjakob) [inline tests/test_classes.cpp:443]: space beteween `)` and `{`.

**Review by wjakob — COMMENTED** [review]
This looks good. Just a few minor change requests.

File: tests/test_classes.cpp (hunk 2 above)
Comment (tjstum) [reply tests/test_classes.cpp]: Sure! I was thinking it was more "defer to someone else." I'll change to `HashWrapper`

**Comment by tjstum** [comment]
Hi @wjakob!

I saw that the test case here fails on PyPy. I am just curious if you know of some reason why it would fail like this (but pass on all CPythons and ... the Windows PyPy?). I can do more investigation if you don't have any leads.

Thanks.

**Comment by wjakob** [comment]
This looks to me like a PyPy issue. The efficient workflow to get it fixed (like with CPython) would be to create a standalone C-API reproducer and submit it here: https://foss.heptapod.net/pypy/pypy/-/issues

**Comment by wjakob** [comment]
It's interesting that the hex encoding of these values is quite similar:
```
>>> hex(114184800867133550)
'0x195aa78ae6f286e'
>>> hex(4725870819294521452)
'0x4195aa78ae6f286c'
```

Makes me wonder if it could be a integer/sign conversion related issue.

**Comment by wjakob** [comment]
Note: in addition to the issue with PyPy, there is a potential overlap with a `hash` function defined in `nanobind/operators.h`. It would be good to test your changes in the context of a C++ file that includes this header. (Maybe it's fine since that other version targets the `nb::self` placeholder value).

**Comment by tjstum** [comment]
Thanks. I'll be sure to test with that header included as well.

I'm traveling for a holiday/work right now, so it might be a few more days/week before I make progress with this effort.

**Comment by wjakob** [comment]
I will close this PR for now. It will reopen it if there are any news (e.g. tracking down those differences in PyPy).

**Pull request closed without merging**